### PR TITLE
[#4581]feat(iceberg) support Gravitino-based multi catalog support for Gravitino Iceberg REST server

### PR DESCRIPTION
**Pull request by theoryxu**
### What changes were proposed in this pull request?

support Gravitino-based multi catalog support for Gravitino Iceberg REST server

### Why are the changes needed?

fixes: #4581

### Does this PR introduce _any_ user-facing change?

add a property key

### How was this patch tested?

1. add UT
2. manual test

### CODE REVIEW

**Comment by theoryxu** [comment]
@jerryshao @FANNG1 please help to review, Thanks.

File: iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * This provider proxy Gravitino lakehouse-iceberg catalogs.
+ *
+ * <p>For example, there are one catalog named iceberg_catalog in metalake
+ *
+ * <p>The catalogName is iceberg_catalog
+ */
+public class GravitinoBasedIcebergTableOpsProvider implements IcebergTableOpsProvider {
```
Comment (FANNG1) [inline iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java:42]: after adding `GravitinoBasedIcebergTableOpsProvider`, How about add cache expire time in `IcebergTableOpsManager`? to refresh Iceberg catalogs after altering.

Comment (FANNG1) [inline iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java:42]: Seems more proper to place xxxTableOpsProvder in `iceberg-rest-server` module, WDYT?

Comment (theoryxu) [reply iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java]: If we do that, IcebergTableOpsManager also needs be replaced togther.

File: iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  private final AtomicReference<GravitinoMetalake> metalakeWrapper = new AtomicReference<>();
+
+  private String gravitinoUri;
+
+  private String gravitinoMetalake;
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    this.gravitinoUri = properties.get(IcebergConstants.GRAVITINO_URI);
+    this.gravitinoMetalake = properties.get(IcebergConstants.GRAVITINO_METALAKE);
```
[the remark this answers is not in the document]
Comment (theoryxu) [reply iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/ops/GravitinoBasedIcebergTableOpsProvider.java]: done

**Comment by theoryxu** [comment]
@FANNG1 All comments have been resolved. Could you review it again? Thanks

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java [hunk 3]
```diff
@@ -35,16 +43,41 @@ public class IcebergTableOpsManager implements AutoCloseable {
...
+    this.icebergTableOpsCache =
+        Caffeine.newBuilder()
+            .expireAfterWrite(
+                (new IcebergConfig(properties))
+                    .get(IcebergConfig.ICEBERG_REST_CATALOG_CACHE_EVICTION_INTERVAL),
+                TimeUnit.MILLISECONDS)
+            .removalListener(
+                (k, v, c) -> {
+                  LOG.info("Remove IcebergTableOps cache {}.", k);
+                  try {
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java:66]: We could make GravitinoCatalogProvider implement `Closable`, and close it in remove listener
```java
if (v instanceOf Closeable) {
  ((Closable)v).close()
}
```

File: iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/IcebergConfig.java [hunk 4]
```diff
@@ -166,6 +166,13 @@ public class IcebergConfig extends Config implements OverwriteDefaultConfig {
           .stringConf()
           .create();
 
+  public static final ConfigEntry<Long> ICEBERG_REST_CATALOG_CACHE_EVICTION_INTERVAL =
+      new ConfigBuilder(IcebergConstants.ICEBERG_REST_CATALOG_CACHE_EVICTION_INTERVAL)
+          .doc("Catalog cache eviction interval.")
+          .version(ConfigConstants.VERSION_0_7_0)
```
Comment (FANNG1) [inline iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/IcebergConfig.java:172]: default to 1 hour?

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    Preconditions.checkArgument(
+        "lakehouse-iceberg".equals(catalog.provider()),
+        String.format("%s.%s is not iceberg catalog", getMetalake(), catalogName));
+
+    Map<String, String> properties =
+        IcebergPropertiesUtils.toIcebergCatalogProperties(catalog.properties());
+    return new IcebergTableOps(new IcebergConfig(properties));
+  }
+
+  public GravitinoMetalake getMetalake() {
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java:89]: close it in `close()` method?

**Comment by FANNG1** [comment]
LGTM except minor comments

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java (hunk 5 above)
Comment (theoryxu) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java]: `GravitinoMetalake` doesn't have `close()`, Only `GravitinoAdminClient` has it. I have rebuilt this part.

File: iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/IcebergConfig.java (hunk 4 above)
Comment (theoryxu) [reply iceberg/iceberg-common/src/main/java/org/apache/gravitino/iceberg/common/IcebergConfig.java:172]: done

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java (hunk 3 above)
Comment (theoryxu) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java]: I modified GravitinoCatalogProvider to maintain a GravitinoAdminClient and ensure it closes when IcebergTableOpsManager closes.

File: docs/iceberg-rest-service.md [hunk 6]
```diff
@@ -215,6 +217,24 @@ You can access different catalogs by setting the `prefix` to the specific catalo
 ...
 ```
 
+##### gravitino-based catalog provider
```
Comment (FANNG1) [inline docs/iceberg-rest-service.md:220]: Gravitino based?

File: docs/iceberg-rest-service.md [hunk 7]
```diff
@@ -178,6 +178,8 @@ The Gravitino Iceberg REST server supports multiple catalogs and offers a config
 |----------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------|----------|---------------|
 | `gravitino.iceberg-rest.catalog-provider`    | Catalog provider class name, you can develop a class that implements `IcebergTableOpsProvider` and add the corresponding jar file to the Iceberg REST service classpath directory.   | `config-based-provider`     | No       | 0.7.0         |
 
+##### config-based catalog provider
```
Comment (FANNG1) [inline docs/iceberg-rest-service.md:181]: Configuration based?

Comment (theoryxu) [reply docs/iceberg-rest-service.md]: fixed

File: docs/iceberg-rest-service.md (hunk 6 above)
Comment (theoryxu) [reply docs/iceberg-rest-service.md]: fixed

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java [hunk 8]
```diff
@@ -97,8 +126,20 @@ private String shelling(String rawPrefix) {
     }
   }
 
+  private void closeIcebergTableOps(IcebergTableOps ops) {
+    try {
+      ops.close();
+    } catch (Exception ex) {
+      LOG.warn("close iceberg table ops fail: {}", ops);
+      throw new RuntimeException(ex);
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java:134]: what will happen if throwing exception here? I prefer to just print warning logs.

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  @VisibleForTesting
+  void setClient(GravitinoAdminClient client) {
+    this.client = client;
+  }
+
+  @Override
+  public void close() throws Exception {
+    client.close();
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java:94]: check whether the client is null?

Comment (theoryxu) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java:94]: ok

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java (hunk 8 above)
Comment (theoryxu) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java:134]: ok

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        String.format("%s.%s is not iceberg catalog", gravitinoMetalake, catalogName));
+
+    Map<String, String> properties =
+        IcebergPropertiesUtils.toIcebergCatalogProperties(catalog.properties());
+    return new IcebergTableOps(new IcebergConfig(properties));
+  }
+
+  @VisibleForTesting
+  void setClient(GravitinoAdminClient client) {
+    this.client = client;
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/provider/GravitinoBasedIcebergTableOpsProvider.java:89]: please remove this

File: iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java [hunk 11]
```diff
@@ -97,8 +126,19 @@ private String shelling(String rawPrefix) {
     }
   }
 
+  private void closeIcebergTableOps(IcebergTableOps ops) {
+    try {
+      ops.close();
+    } catch (Exception ex) {
+      LOG.warn("close iceberg table ops fail: {}, {}", ops, ex);
```
Comment (FANNG1) [inline iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java:133]: close -> Close

Comment (FANNG1) [reply iceberg/iceberg-rest-server/src/main/java/org/apache/gravitino/iceberg/service/IcebergTableOpsManager.java:133]: iceberg -> Iceberg

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@theoryxu , thanks for your contribution!

**Pull request merged**